google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -146,8 +145,18 @@ def stop
             end
           end
           stop_pool.map(&:join)
-          # Stop the buffer after the streams are all stopped
-          synchronize { @buffer.stop }
+
+          wait_pool = synchronize do
+            @stream_pool.map do |stream|
+              Thread.new { stream.wait! }
+            end
+          end
+
+          @final_stopped_thread = Thread.new do
+            wait_pool.map(&:join)
+            # Shutdown the buffer TimerTask (and flush the buffer) after the streams are all stopped.
+            synchronize { @buffer.stop }
+          end
```

Comment:
Looking at this code now, I wonder if we really need separate threads for every stream. Maybe we can just run a single thread that waits on every stream sequentially (since all we're doing is waiting for termination, which is all happening in parallel anyway). And then that same thread can call `@buffer.stop` at the end. So we only have to spawn one thread rather than N+1 threads.

Reply:
Sounds good, I'll take a look.

Reply:
Updated. I added temporary debug logging and boosted the number of streams in the two acceptance tests to manually verify that the sequential calls work OK. The sequential `stop` calls all return without noticeable delay. The sequential `wait!` calls also do, except in a single "blocking" call for 1 of the 2 sleeping callbacks. Presumably the other sleeping callback has finished executing in parallel about the time `wait!` is called for its stream.

Reply:
Sounds good, thanks for checking!